### PR TITLE
[16.0][MIG] contract_sale_generation: Migration to 16.0

### PR DESCRIPTION
**Pull request by JesusValdez96**
Replacing https://github.com/OCA/contract/pull/914
- Migration according to migration guide.
- Fix tests failing due to dates and discount field not showing if user doesn't have group_discount_per_so_line

### CODE REVIEW

**Comment by JesusValdez96** [comment]
@rousseldenis could you review this?

File: contract_sale_generation/models/contract.py [hunk 1]
```diff
@@ -32,8 +32,12 @@ def _prepare_sale(self, date_ref):
             sale.payment_term_id = self.payment_term_id.id
         if self.fiscal_position_id:
             sale.fiscal_position_id = self.fiscal_position_id.id
-        # Get other sale values from partner onchange
-        sale.onchange_partner_id()
+        # Get other sale values
+        sale._compute_partner_shipping_id()
```
Comment (rousseldenis) [inline contract_sale_generation/models/contract.py]: Should we really call the computes ? Wouldn't the values automatically computed ?

File: contract_sale_generation/models/contract_line.py [hunk 2]
```diff
@@ -11,7 +11,7 @@ def _prepare_sale_line_vals(self, dates, order_id=False):
         sale_line_vals = {
             "product_id": self.product_id.id,
             "product_uom_qty": self._get_quantity_to_invoice(*dates),
-            "product_uom": self.uom_id.id,
+            "product_uom": self.product_id.uom_id.id,
```
Comment (rousseldenis) [inline contract_sale_generation/models/contract_line.py]: Why changing this ? If we want to use a particular uom on contract line, that one should be used.

File: contract_sale_generation/tests/test_contract_sale.py [hunk 3]
```diff
@@ -87,7 +88,7 @@ def test_contract_count_sale_2(self):
         self.assertEqual(set(action["domain"][0][2]), set(orders.ids))
 
     def test_cron_recurring_create_sale(self):
-        self.contract_line.date_start = "2020-01-01"
+        self.contract_line.date_start = "2020-01-15"
```
Comment (rousseldenis) [inline contract_sale_generation/tests/test_contract_sale.py]: Why this change ?

**Review by rousseldenis — COMMENTED** [review]
Some questions

**Comment by JesusValdez96** [comment]
Thanks @rousseldenis those changes were certainly not necessary.
I made some changes, could you review again?

**Review by rousseldenis — APPROVED** [review]

**Review by hugho-ad — APPROVED** [review]
